src/devices/sound/ymfm.cpp

```diff
@@ -0,0 +1,1658 @@
+{
+	if (rawrate == 0)
+		return 0;
+	u8 rate = rawrate * 2 + (keycode >> (m_regs.ksr() ^ 3));
+	return (rate < 64) ? rate : 63;
+}
+
+
+//-------------------------------------------------
+//  start_attack - start the attack phase; called
+//  when a keyon happens or when an SSG-EG cycle
+//  is complete and restarts
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_operator<RegisterType>::start_attack(u8 keycode)
+{
+	// don't change anything if already in attack state
+	if (m_env_state == ENV_ATTACK)
+		return;
+	m_env_state = ENV_ATTACK;
+
+	// generally not inverted at start, except if SSG-EG is
+	// enabled and one of the inverted modes is specified
+	m_ssg_inverted = m_regs.ssg_eg_enabled() & BIT(m_regs.ssg_eg_mode(), 2);
+
+	// reset the phase when we start an attack
+	m_phase = 0;
+
+	// if the attack rate >= 62 then immediately go to max attenuation
+	if (effective_rate(m_regs.attack_rate(), keycode) >= 62)
+		m_env_attenuation = 0;
+
+	// log keys
+	if (LOG_KEY_ONS)
+	{
+		printf("KeyOn %d.%d: freq=%04X dt2=%d fb=%d alg=%d dt=%d mul=%X tl=%02X ksr=%d adsr=%02X/%02X/%02X/%X sl=%X pan=%c%c",
+			m_regs.chnum(), m_regs.opnum(),
+			m_regs.block_freq(),
+			m_regs.detune2(),
+			m_regs.feedback(),
+			m_regs.algorithm(),
+			m_regs.detune(),
+			m_regs.multiple(),
+			m_regs.total_level(),
+			m_regs.ksr(),
+			m_regs.attack_rate(),
+			m_regs.decay_rate(),
+			m_regs.sustain_rate(),
+			m_regs.release_rate(),
+			m_regs.sustain_level(),
+			m_regs.pan_left() ? 'L' : '-',
+			m_regs.pan_right() ? 'R' : '-');
+		if (m_regs.ssg_eg_enabled())
+			printf(" ssg=%X", m_regs.ssg_eg_mode());
+		if (m_regs.lfo_enabled() && ((m_regs.lfo_am_enabled() && m_regs.lfo_am_sensitivity() != 0) || m_regs.lfo_pm_sensitivity() != 0))
+			printf(" am=%d pm=%d w=%d", m_regs.lfo_am_enabled() ? m_regs.lfo_am_sensitivity() : 0, m_regs.lfo_pm_sensitivity(), m_regs.lfo_waveform());
+		if (m_regs.noise_enabled() && m_regs.opnum() == 3 && m_regs.chnum() == 7)
+			printf(" noise=1");
+		printf("\n");
+	}
+}
+
+
+//-------------------------------------------------
+//  start_release - start the release phase;
+//  called when a keyoff happens
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_operator<RegisterType>::start_release()
+{
+	// don't change anything if already in release state
+	if (m_env_state == ENV_RELEASE)
+		return;
+	m_env_state = ENV_RELEASE;
+
+	// adjust attenuation if inverted due to SSG-EG
+	if (m_ssg_inverted)
+		m_env_attenuation = 0x200 - m_env_attenuation;
+}
+
+
+//-------------------------------------------------
+//  clock_keystate - clock the keystate to match
+//  the incoming keystate
+//-------------------------------------------------
+
+template<class RegisterType>
+void ymfm_operator<RegisterType>::clock_keystate(u8 keystate, u8 keycode)
+{
+	// has the key changed?
+	if ((keystate ^ m_key_state) != 0)
```

Comment:
This isn’t necessarily going to work if the `keystate` is something other than 1 or 0.

Reply:
It is always, but I will add an assert to ensure it.